File: python/src/hotel_queries.py
import sys
import math

class SegmentTree:

    @staticmethod
    def parent(i):
        return (i - 1) // 2

    @staticmethod
    def left(i):
        return i * 2 + 1

    @staticmethod
    def right(i):
        return i * 2 + 2

    @staticmethod
    def sibling(i):
        if i % 2 == 0:
            return i - 1
        else:
            return i + 1

    # should only be called using an index from the last level
    def original_index(self, tree_index):
        return tree_index - (2**(self.n - 1) - 1) + 1 # for 1 indexed hotels

    def comparison(self, i, j):
        return self.tree[i] > self.tree[j]

    def __init__(self, values) -> None:
        self.n = math.ceil(math.log2(len(values))) + 1
        self.tree = [0 for i in range(2**self.n - 1)]

        # fill initial values
        for (i, value) in zip(range(2**(self.n - 1) - 1, 2**self.n - 1), values):
            self.tree[i] = value

        for i in range(2**self.n - 1 - 1, 0, -2):
            if self.comparison(i, i - 1):
                self.tree[self.parent(i)] = self.tree[i]
            else:
                self.tree[self.parent(i)] = self.tree[i - 1]

    def search(self, value):
        index = 0

        if value > self.tree[index]:
            return None

        while index < (2**(self.n - 1) - 1):

            if value <= self.tree[self.left(index)]:
                index = self.left(index)
                continue

            if value <= self.tree[self.right(index)]:
                index = self.right(index)
                continue

        return (index, self.tree[index])

    def update_value(self, tree_index, value):
        self.tree[tree_index] = value

        while tree_index != 0:
            sibling_index = self.sibling(tree_index)

            if self.comparison(tree_index, sibling_index):
                self.tree[self.parent(tree_index)] = self.tree[tree_index]
            else:
                self.tree[self.parent(tree_index)] = self.tree[sibling_index]

            tree_index = self.parent(tree_index)

    def return_original_values(self):
        return self.tree[2**(self.n - 1) - 1, 2**self.n - 1]

if __name__ == "__main__":

    first_line = sys.stdin.readline().split(" ")
    n_hotels = int(first_line[0])
    m_groups = int(first_line[1])

    hotel_capacity = list(map(int, sys.stdin.readline().split(" ")))
    group_rooms = list(map(int, sys.stdin.readline().split(" ")))

    tree = SegmentTree(hotel_capacity)

    for rooms in group_rooms:
        response = tree.search(rooms)

        if response:
            (tree_index, current_rooms) = response
            tree.update_value(tree_index, current_rooms - rooms)
            print(f"{tree.original_index(tree_index)}", end=" ")
        else:
            print("0", end=" ")

File: rust/src/hotel_queries.rs
use std::io::{stdin, Read};

struct SegmentTree {
    base_index: usize,
    tree: Vec<usize>,
}

impl SegmentTree {
    fn parent(i: usize) -> usize {
        return (i - 1) / 2;
    }

    fn left(i: usize) -> usize {
        return i * 2 + 1;
    }

    fn right(i: usize) -> usize {
        return i * 2 + 2;
    }

    fn sibling(i: usize) -> usize {
        if i % 2 == 0 {
            return i - 1;
        } else {
            return i + 1;
        }
    }

    fn original_index(&self, tree_index: usize) -> usize {
        return tree_index - self.base_index + 1;
    }

    fn comparison(&self, i: usize, j: usize) -> bool {
        return self.tree[i] > self.tree[j];
    }

    fn new(values: Vec<usize>) -> Self {
        let levels: usize = f64::ceil(f64::log2(values.len() as f64)).round() as usize + 1;
        let base_index = (2 as usize).pow((levels - 1) as u32) - 1;
        let tree = vec![0; (2 as usize).pow(levels as u32) - 1];

        let mut segment_tree = SegmentTree { base_index, tree };

        segment_tree.treeify(values);
        segment_tree
    }

    fn treeify(&mut self, values: Vec<usize>) {
        for (i, &value) in (self.base_index..(self.base_index * 2 + 1)).zip(values.iter()) {
            self.tree[i] = value;
        }

        for i in (1..(self.base_index * 2 + 1)).step_by(2).rev() {
            if self.comparison(i, i + 1) {
                self.tree[SegmentTree::parent(i)] = self.tree[i];
            } else {
                self.tree[SegmentTree::parent(i)] = self.tree[i + 1];
            }
        }
    }

    fn search(&self, value: usize) -> Option<(usize, usize)> {
        let mut index = 0;
        let mut result: Option<(usize, usize)> = None;

        if value <= self.tree[index] {
            while index < self.base_index {
                let left = SegmentTree::left(index);
                let right = SegmentTree::right(index);

                if value <= self.tree[left] {
                    index = left;
                    continue;
                }

                if value <= self.tree[right] {
                    index = right;
                    continue;
                }
            }

            result = Some((index, self.tree[index]));
        }

        result
    }

    fn update_value(&mut self, index: usize, value: usize) {
        let mut tree_index = index;
        self.tree[tree_index] = value;

        while tree_index > 0 {
            let sibling_index = SegmentTree::sibling(tree_index);
            let parent_index = SegmentTree::parent(tree_index);

            if self.comparison(tree_index, sibling_index) {
                self.tree[parent_index] = self.tree[tree_index];
            } else {
                self.tree[parent_index] = self.tree[sibling_index];
            }

            tree_index = parent_index;
        }
    }
}

fn main() {
    let mut input = String::new();
    stdin().read_to_string(&mut input).unwrap();
    let mut input = input.lines();

    let first_line: Vec<&str> = input.next().unwrap().split(' ').collect();
    let _n_hotels: usize = first_line[0].parse().unwrap();
    let _m_groups: usize = first_line[1].parse().unwrap();

    let hotel_capacity: Vec<usize> = input
        .next()
        .unwrap()
        .split(' ')
        .map(|value| value.parse().unwrap())
        .collect();
    let group_rooms: Vec<usize> = input
        .next()
        .unwrap()
        .split(' ')
        .map(|value| value.parse().unwrap())
        .collect();

    let mut tree = SegmentTree::new(hotel_capacity);

    for &room in group_rooms.iter() {
        if let Some((tree_index, current_rooms)) = tree.search(room) {
            tree.update_value(tree_index, current_rooms - room);
            print!("{} ", tree.original_index(tree_index));
        } else {
            print!("0 ");
        }
    }
}
